File: javascript/03_chapter/BinarySearchST.js
/**
 * 顺序查找（基于有序数组）
 */
class BinarySearchST {
    constructor(){
        this.keys = [];
        this.vals = [];
    }

    print(){
        for (let i = 0 ; i < this.keys.length; i++){
            console.log(this.keys[i]);
        }
    }

    get(key){
        for (let i = 0; i < this.keys.length; i++){
            if (this.keys[i] === key){
                return i;
            }
        }
    }

    rank(key){
        let lo = 0, hi = this.keys.length - 1;
        while (lo <= hi){
            let mid = lo + parseInt((hi - lo) / 2);
            let midKey = this.keys[mid]
            if ( key < midKey) {
                hi = mid - 1;
            } else if ( key > midKey) {
                lo = mid + 1;
            } else {
                return mid;
            }
        }
        return lo;
    }

    put(key, val){
        let i = this.rank(key);
        if (i && this.keys[i] === key){
            this.vals[i] = val;
        } else {
            for (let j = this.vals.length; j > i; j--){
                this.keys[j] = this.keys[j-1];
                this.vals[j] = this.vals[j-1];
            }
            this.keys[i] = key;
            this.vals[i] = val;
        }
    }
}

let st = new BinarySearchST();
st.put('L', 11);
st.put('P', 10);
st.put('M', 9);
st.put('X', 7);
st.put('H', 5);
st.put('C', 4);
st.put('R', 3);
st.put('A', 8);
st.put('E', 12);
st.put('S', 0);

st.print();

console.log(st.get('A'))
console.log(st.get('C'))
console.log(st.get('M'))
console.log(st.get('S'))
console.log(st.get('X'))

File: algs4/03_chapter/BinarySearchST.java
import java.util.NoSuchElementException;

/**
 * 3.1.5 二分查找（基于有序数组）
 * 编译： javac -Xlint:unchecked -cp :../stdlib.jar:../01_chapter BinarySearchST.java
 * 运行： java -cp :../stdlib.jar:../01_chapter BinarySearchST.java < tinyST.txt
 * 结果：
 * L 11
 * P 10
 * M 9
 * X 7
 * H 5
 * C 4
 * R 3
 * A 8
 * E 12
 * S 0
 */

public class BinarySearchST<Key extends Comparable<Key>, Value> {
    private Key[] keys;
    private Value[] vals;
    private int N;

    public BinarySearchST(int capacity){
        keys = (Key[]) new Comparable[capacity];
        vals = (Value[]) new Object[capacity];
    }

    public int size(){
        return N;
    }

    public boolean isEmpty(){
        return N == 0;
    }

    public Value get(Key key){
        if (isEmpty()) return null;
        int i = rank(key);
        if (i < N && keys[i].compareTo(key) == 0){
            return vals[i];
        } else {
            return null;
        }
    }

    // 核心算法（递归）rank: 小于key的数量
    public int rank(Key key){
        return rank(key, 0, N - 1);
    }

    public int rank(Key key, int lo, int hi){
        if (hi < lo) return lo;
        int mid = lo + (hi - lo) / 2;
        int cmp = key.compareTo(keys[mid]);
        if (cmp < 0) {
            return rank(key, lo, mid - 1);
        } else if (cmp > 0){
            return rank(key, mid + 1, hi);
        } else {
            return mid;
        }
    }

    // 核心算法(迭代)
    // public int rank(Key key){
    //     int lo = 0, hi = N - 1;
    //     while(lo <= hi){
    //         int mid = lo + (hi - lo) / 2;
    //         int cmp = key.compareTo(keys[mid]);
    //         if (cmp < 0) hi = mid - 1;
    //         else if (cmp > 0) lo = mid + 1;
    //         else return mid;
    //     }
    //     return lo;
    // }

    public void put(Key key, Value val){
        int i = rank(key);
        if (i < N  && keys[i].compareTo(key) == 0){
            vals[i] = val;
            return;
        }
        for (int j = N; j > i; j--){
            keys[j] = keys[j-1];
            vals[j] = vals[j-1];
        }
        keys[i] = key;
        vals[i] = val;
        N++;
    }

    public Key min(){
        if (isEmpty()) throw new NoSuchElementException("called min() with empty symbol table");
        return keys[0];
    }

    public Key max(){
        if (isEmpty()) throw new NoSuchElementException("called min() with empty symbol table");
        return keys[N-1];
    }

    public boolean contains(Key key){
        if (key == null) throw new IllegalArgumentException("argument to contains() is null");
        return get(key) != null;
    }

    public Iterable<Key> keys(){
        return keys(min(), max());
    }

    public Iterable<Key> keys(Key lo, Key hi){
        if (lo == null) throw new IllegalArgumentException("first argument to keys() is null"); 
        if (hi == null) throw new IllegalArgumentException("second argument to keys() is null"); 

        Queue<Key> queue = new Queue<Key>();
        if (lo.compareTo(hi) > 0) return queue;
        for (int i = rank(lo); i < rank(hi); i++){
            queue.enqueue(keys[i]);
        }
        if (contains(hi)) queue.enqueue(keys[rank(hi)]);
        return queue;
    }

    public static void main(String[] args) { 
        BinarySearchST<String, Integer> st = new BinarySearchST<String, Integer>(100);
        for (int i = 0; !StdIn.isEmpty(); i++) {
            String key = StdIn.readString();
            st.put(key, i);
        }
        for (String s : st.keys())
            StdOut.println(s + " " + st.get(s));
    }
}